addons/account/models/account_move.py

```diff
@@ -4260,10 +4260,16 @@ def _post(self, soft=True):
                 purchase_amls = invoice.line_ids.filtered(lambda line: line.partner_id and line.account_id.account_type == 'liability_payable')
                 for partner in purchase_amls.mapped('partner_id'):
                     supplier_count[partner] += 1
+        # Group partner by count to minimize the call to _increase_rank in order to minimize the number of savepoint
+        count_per_customer, count_per_supplier = defaultdict(self.env['res.partner'].browse), defaultdict(self.env['res.partner'].browse)
```

Comment:
Maybe we can use something like
```python
count2customers = self.filtered(lambda m: m.is_sale_document()).partner_id.groupby(customer_count.get)
```
To stay in linear complexity?

Reply:
It's already linear AFAIK. And partner_id.grouby does not seems to exist and the customer count is not limited to only sale_document as you can see just above.

Reply:
`+=` on recordsets is linear because it is immutable, then doing `+=` in a loop is `O(n²)`

Reply:
Indeed, My bad. 
But I think if we have a perf issue with the size of the customer_count, we will have other perf issue with the rest of the transaction as well. Anyway

We cannot get the partner list from to_post we have to based it on customer_count so I guess I would give us

```suggestion
        count_per_customer = self.env['res.partner'].browse(p.id for p in customer_count.keys()).grouped(customer_count.get)
        count_per_supplier = self.env['res.partner'].browse(p.id for p in supplier_count.keys()).grouped(supplier_count.get)
```